### PR TITLE
Grad function Conv2d

### PR DESCRIPTION
**Pull request by aanurraj**
## Description
This PR implements grad functions mainly conv2d, which in the process requires conv_transpose2d.

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/sympc/tensor/grads/grad_functions.py [hunk 1]
```diff
@@ -432,6 +435,159 @@ def backward(ctx: Dict[str, Any], grad: MPCTensor) -> MPCTensor:
...
+            weight_size,
+            dilation,
+            grad.session,
+        ]
+        args = [[el] + common_args for el in grad.share_ptrs]
+
+        shares = parallel_execution(GradConv2d.function_aux, grad.session.parties)(args)
+        grad_input_padding = MPCTensor(shares=shares, session=grad.session)
+
+        output_padding_tensor = grad_input_padding.reconstruct()
```
Comment (gmuraru) [inline src/sympc/tensor/grads/grad_functions.py:542]: Q: Does this have the correct value?

File: src/sympc/tensor/grads/grad_functions.py [hunk 2]
```diff
@@ -432,6 +435,159 @@ def backward(ctx: Dict[str, Any], grad: MPCTensor) -> MPCTensor:
...
+        output_padding = tuple(output_padding_tensor.to(torch.int).tolist())
+
+        input_grad = grad.conv_transpose2d(
```
Comment (gmuraru) [inline src/sympc/tensor/grads/grad_functions.py:545]: Q: Does this?

File: src/sympc/tensor/grads/grad_functions.py (hunk 1 above)
Comment (aanurraj) [reply src/sympc/tensor/grads/grad_functions.py:542]: yes this gives correct value

File: src/sympc/tensor/grads/grad_functions.py (hunk 2 above)
Comment (aanurraj) [reply src/sympc/tensor/grads/grad_functions.py:545]: No if the parties > 2 we get the over-scaled value we talked about.

File: src/sympc/api.py [hunk 3]
```diff
@@ -36,6 +36,12 @@
     ("sympc.protocol.spdz", protocol.spdz),
     ("sympc.protocol.spdz.spdz", protocol.spdz.spdz),
     ("sympc.utils", utils),
+    ("sympc.tensor.grads", tensor.grads),
```
Comment (gmuraru) [inline src/sympc/api.py:39]: Q: Do we need those here?

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 4]
```diff
@@ -159,11 +159,12 @@ def test_ops_mpc_mpc_div(get_clients, nr_clients, op_str) -> None:
     assert np.allclose(result, expected_result, rtol=10e-4)
 
 
-@pytest.mark.parametrize("nr_clients", [2])
+@pytest.mark.parametrize("nr_clients", [2, 3, 4])
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:162]: :+1:

File: src/sympc/tensor/grads/grad_functions.py [hunk 5]
```diff
@@ -432,6 +435,160 @@ def backward(ctx: Dict[str, Any], grad: MPCTensor) -> MPCTensor:
         return grad.reshape(shape)
 
 
+class GradConv2d(GradFunc):
+    """The multiplication gradient function."""
+
+    @staticmethod
+    def function_aux(
```
Comment (gmuraru) [inline src/sympc/tensor/grads/grad_functions.py:442]: Q: Could we rename this? Maybe to something like `get_grad_input_padding`?

File: src/sympc/tensor/grads/__init__.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 """Used to keep track of what gradient functions we have implemented."""
 
 from .grad_functions import GRAD_FUNCS
+from .grad_functions import GradConv2d
```
Comment (gmuraru) [inline src/sympc/tensor/grads/__init__.py:4]: Add this to `__all__` down bellow

File: src/sympc/tensor/grads/grad_functions.py (hunk 5 above)
Comment (gmuraru) [reply src/sympc/tensor/grads/grad_functions.py:442]: Could we also add a test (in the `grad_test` py file) only for this method?

**Review by gmuraru — APPROVED** [review]

File: src/sympc/tensor/grads/grad_functions.py (hunk 5 above)
Comment (aanurraj) [reply src/sympc/tensor/grads/grad_functions.py:442]: sure :D

File: src/sympc/tensor/grads/__init__.py (hunk 6 above)
Comment (aanurraj) [reply src/sympc/tensor/grads/__init__.py:4]: done

**Review by gmuraru — APPROVED** [review]
LGTM!

**Pull request merged**